### ISSUE TITLE
font size in wxMM_METRIC wxPaintDC

### ISSUE THREAD

**Issue by EddieM0710** [wxWidgets/wxWidgets#24176]
I am drawing text and rectangles into a wxPaintDC using wxMM_Metric memory mapped mode. It appears that the text size is being drawn larger than it should be.  I.e., if I use a 6 point font the dc.GetTextExtent function says the height is 10mm. It should be about 2mm.
I have a work around for my app (i.e., scaling the font to what looks right) but this does seem like a bug.
Here is a sample screen clip and the sample code that generated it.
![Screenshot](https://github.com/wxWidgets/wxWidgets/assets/29614350/e39337ad-68b3-443f-b51d-36daaa929242)
`
#include "wx/wxprec.h"

#ifndef WX_PRECOMP
#include "wx/wx.h"
#endif

#include <vector>
#include <wx/panel.h>
#include <wx/event.h>

#include <wx/display.h>

#define ID_ZOOMFRAME 10000
#define ID_SCROLLEDWINDOW 10005

#define SYMBOL_ZOOMFRAME_STYLE wxCAPTION|wxRESIZE_BORDER|wxSYSTEM_MENU|wxCLOSE_BOX
#define SYMBOL_ZOOMFRAME_TITLE _("ZoomFrame")
#define SYMBOL_ZOOMFRAME_IDNAME ID_ZOOMFRAME
#define SYMBOL_ZOOMFRAME_SIZE wxSize(1000, 1000)
#define SYMBOL_ZOOMFRAME_POSITION wxDefaultPosition

static const char* mondrian_xpm[ ] = {
    /* columns rows colors chars-per-pixel */
    "32 32 6 1",
    "  c Black",
    ". c Blue",
    "X c #00bf00",
    "o c Red",
    "O c Yellow",
    "+ c Gray100",
    /* pixels */
    "                                ",
    " oooooo +++++++++++++++++++++++ ",
    " oooooo +++++++++++++++++++++++ ",
    " oooooo +++++++++++++++++++++++ ",
    " oooooo +++++++++++++++++++++++ ",
    " oooooo +++++++++++++++++++++++ ",
    " oooooo +++++++++++++++++++++++ ",
    " oooooo +++++++++++++++++++++++ ",
    "                                ",
    " ++++++ ++++++++++++++++++ .... ",
    " ++++++ ++++++++++++++++++ .... ",
    " ++++++ ++++++++++++++++++ .... ",
    " ++++++ ++++++++++++++++++ .... ",
    " ++++++ ++++++++++++++++++ .... ",
    " ++++++ ++++++++++++++++++      ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++ ++++++++++++++++++ ++++ ",
    " ++++++                    ++++ ",
    " ++++++ OOOOOOOOOOOO XXXXX ++++ ",
    " ++++++ OOOOOOOOOOOO XXXXX ++++ ",
    " ++++++ OOOOOOOOOOOO XXXXX ++++ ",
    " ++++++ OOOOOOOOOOOO XXXXX ++++ ",
    " ++++++ OOOOOOOOOOOO XXXXX ++++ ",
    " ++++++ OOOOOOOOOOOO XXXXX ++++ ",
    "                                "
};
//------------

class MyApp : public wxApp
{
public:
    virtual bool OnInit( );
};

//------------

DECLARE_APP( MyApp )
IMPLEMENT_APP( MyApp )

//------------

class MyFrame : public wxFrame
{
public:

    MyFrame( const wxString& title );

    void CreateControls( );

private:

    wxScrolledWindow* itemScrolledWindow1;

    DECLARE_EVENT_TABLE( )
};

//------------

BEGIN_EVENT_TABLE( MyFrame, wxFrame )
END_EVENT_TABLE( )

//------------

class MyCanvas : public wxScrolled<wxWindow>
{
public:
    MyCanvas( wxWindow* parent )
        : wxScrolled<wxWindow>( parent, wxID_ANY )
    {
        frame = parent;
        SetScrollRate( 10, 10 );
        SetVirtualSize( 1000, 1000 );
        SetBackgroundColour( *wxWHITE );
        SetScrollbars( 3, 3, 1000, 1000, 0, 0 );
        EnableScrolling( true, true );

        Bind( wxEVT_PAINT, &MyCanvas::OnPaint, this );
    }

    void OnPaint( wxPaintEvent& event );

private:
    wxWindow* frame;
};

//------------

//------------

bool MyApp::OnInit( )
{
    MyFrame* frame = new MyFrame( wxT( "Minimal wxWidgets App" ) );
    frame->Show( true );
    return true;
}

//------------

//------------

MyFrame::MyFrame( const wxString& title )
    : wxFrame( NULL, wxID_ANY, title )
{
    SetIcon( wxIcon( mondrian_xpm ) );

    CreateControls( );
}

//------------

void MyFrame::CreateControls( )
{
    MyFrame* itemFrame1 = this;

    wxBoxSizer* frameVBoxSizer = new wxBoxSizer( wxVERTICAL );
    itemFrame1->SetSizer( frameVBoxSizer );

    wxBoxSizer* sliderHBoxSizer = new wxBoxSizer( wxHORIZONTAL );
    frameVBoxSizer->Add( sliderHBoxSizer, 0, wxGROW | wxALL, 5 );

    wxBoxSizer* scrolledWindowHBoxSizer = new wxBoxSizer( wxHORIZONTAL );
    frameVBoxSizer->Add( scrolledWindowHBoxSizer, 1, wxGROW | wxALL, 5 );
    MyCanvas* scrolledWindow = new MyCanvas( itemFrame1 );

    scrolledWindowHBoxSizer->Add( scrolledWindow, 1, wxGROW | wxALL, 5 );

};

//------------

void MyCanvas::OnPaint( wxPaintEvent& event )
{
    wxPaintDC dc( this );
    DoPrepareDC( dc );
    dc.SetMapMode( wxMM_METRIC );
    dc.Clear( );

    wxFont font = *wxSWISS_FONT;
    font.SetPointSize( 6 );
    // wxSize txtExt = dc.GetTextExtent( "Some Text" );
    dc.SetFont( font );
    dc.SetPen( *wxRED_PEN );
    dc.SetBrush( *wxWHITE_BRUSH );
    dc.DrawRectangle( wxRect( 0, 0, 10 * 25.4, 10 * 25.4 ) );

    dc.SetPen( *wxGREEN_PEN );
    wxSize txtExt = dc.GetTextExtent( "Some Text" );
    wxRect txtRect = wxRect( 2, 2, txtExt.x, txtExt.y );
    dc.DrawRectangle( txtRect );

    dc.SetPen( *wxBLUE_PEN );

    wxRect blueRect = wxRect( 60, 2, 25.4, 25.4 );
    dc.DrawRectangle( blueRect );

    dc.DrawLabel( "Some Text", txtRect, wxALIGN_CENTER );

    wxString line;

    line.Printf( "The Red Rectangle is 10in X 10in." );
    dc.DrawText( line, 2, txtExt.y + 10 );

    line.Printf( "The Green Rectangle is %dmm x %dmm with 6 point font text.", txtExt.x, txtExt.y );
    dc.DrawText( line, 2, txtExt.y + 20 );

    line.Printf( "The Blue Rectangle is 25.4mm x 25.4mm (1in x 1in).", txtExt.x, txtExt.y );
    dc.DrawText( line, 2, txtExt.y + 30 );

    dc.DrawText( "The point size is defined as 1/72 of the Anglo-Saxon inch (25.4 mm):", 2, txtExt.y + 40 );
    dc.DrawText( "It is approximately 0.0139 inch or 352.8 um.", 2, txtExt.y + 50 );

    line.Printf( "A 6 point font should be about 2.1mm in height.", txtExt.x, txtExt.y );
    dc.DrawText( line, 2, txtExt.y + 60 );

    wxDisplay display;
    wxSize displayPPI = display.GetPPI( );
    line.Printf( "The display PPI is %d x %d.", displayPPI.x, displayPPI.y );
    dc.DrawText( line, 2, txtExt.y + 70 );

    line.Printf( "This is in a wxPaintDC with wxMM_METRIC\n Memory map mode and 6 point Swiss Font" );
    dc.DrawLabel( line, wxRect( 2, txtExt.y + 80, 25.4 * 6, 25 ), wxALIGN_LEFT );

}

//------------
`

**Comment by EddieM0710** [comment]
I’m running on Linux mint 21 with wxwidgets 3.2.3.
The moderator on the wxwidgets forum said it ran the same on Windows. 
[https://forums.wxwidgets.org/viewtopic.php?p=220589#p220589](url)

**Comment by vadz** [comment]
I didn't look at this in details but I guess it's yet another problem due to "physical size" of display not actually being its physical size, but of actually being what it would be if the DPI used corresponded to it. E.g. the size of my 192 DPI display with 3840x2400 resolution is 508x318mm which is 20x12.5in which is exactly what it should be to give this DPI -- but I doubt it's the actual physical size of it.

IOW it looks fine as far as the program can say, it's just that your 96 DPI monitor doesn't actually show 96px per inch...

**Comment by paulcor** [comment]
I believe this is about a `wxFont`, with a size specified in **points**, not being drawn at that point size because `wxMM_METRIC` has scaled the `wxDC`.

**Comment by vadz** [comment]
I think it's correct that font size is affected by scaling, would you really expect everything else to change size but the font remain as it is?

**Comment by paulcor** [comment]
I'll answer your question with another question. What size would you expect `"Text2"` to be here?
```c++
dc.SetMapMode(wxMM_METRIC);
dc.SetFont(wxFont(wxFontInfo(wxSize(0, 10))));  // 10mm height
dc.DrawText("Text1", 5, 5);
dc.SetFont(wxFont(wxFontInfo(10 / 25.4 * 72)));  // 10mm in points
dc.DrawText("Text2", 5, 5 + 10);
```
BTW, nothing says the map mode just sets the logical scale. I had assumed it was a separate mechanism.

**Comment by EddieM0710** [comment]
I’m not real sure what y’all just said but if I’m drawing into a Metric mapped dc I expect the font to be drawn the correct size “relative” to the metric objects drawn regardless of scaling. This looks like a real bug to me. 
If it is not a bug then I need to know how to use the api to make it look right.
If there is a specific size I need to scale my desired font size to make it “look” right then I can live with that. It would be nice to know how to calculate the scale factor rather than picking something that looks close. (But for my app close is all that matters as long as my eventual printed result looks the same.)
If this is a function of physical display monitors then my app is of no value to anyone else unless they have this same display I do. I have real problems with that. The point of the cross-platform library is to hide all the underlying os/equip details from the code. I’ve been using wxWidgets on and off for 20 years and find it hard to believe that is the case.
Eddie

**Comment by vadz** [comment]
Sorry, I feel completely lost here and I probably shouldn't have even posted without looking at this more carefully, so please forget that I did. To say something useful I'd need to (at least) have the answers to the following questions:

- What exactly is expected to happen here?
- What did 64d1d27da912e33423842101b324a52a030a4240 change?
- Is the current behaviour the same under all platforms (original comment implied that it was, the latest one implies it is not) before/after the commit above?
- What is map mode if not exactly the same as changing the scale?

If anybody already has answers to them, they'd be welcome, TIA!

**Comment by doublemax** [comment]
> * What exactly is expected to happen here?
> * What is map mode if not exactly the same as changing the scale?
> 
Personally i would have expected that map mode only changes coordinates, e.g. when drawing a rectangle or similar. But not the font size. Maybe i want to work in millimeters, because i want to layout content on a A4 page and "thinking" in mm is easier for me. Then, when i draw a 30x30 unit rectangle, it should be 3cm on the output device. If i set a pen to 3mm i expect a 3mm line. And that's what it does.

If wxFont was integral part of wxDC, and its size was given in "device units" there would be no ambiguity here. So the question is how the arguments to wxFont::SetPixelSize() and wxFont::SetPointSize() should be interpreted. One could argue that SetPixelSize() does really use device units instead of pixels, then it would match the current behavior.

But when using wxFont::SetPointSize(), which is an absolute physical size, i don't see any justification for overriding he meaning.

Having that said, i'm pretty sure it would be non-trivial to change the current behavior, and it would probably break a lot of existing code. So maybe we should just make sure that the behavior is consistent across platforms, and extend/clarify the wxDC::SetMapMode() documentation.

**Comment by EddieM0710** [comment]
Well, like I said, it looks like a bug to me. However, whether it is a bug or not the docs should reflect the actual behavior. It had crossed my mind (that’s not a long trip) that a lot of existing code would be adversly affected so the best solution, at least until the next major upgrade, is to update docs to reflect that text font drawing doesn’t match the mapped mode. As long as my code is consistent across platforms I’m ok with applying a fixed scale factor.
Eddie

**Comment by vadz** [comment]
> Well, like I said, it looks like a bug to me.

I understand this but I don't understand what exactly do you expect to happen instead.

> However, whether it is a bug or not the docs should reflect the actual behavior. It had crossed my mind (that’s not a long trip) that a lot of existing code would be adversly affected

If the existing behaviour is the same on all platforms changing it is indeed out of the question to change it. But if the behaviour varies across them we should consider bringing them in sync.

**Comment by EddieM0710** [comment]
What I expect is that if I am using a metric mapped dc then when I draw text with a given font size then it would be correct relative to the other objects drawn. I.e a 6 point font should be about 2.1mm (that’s what an 6point font should be) and look like that relative to a 2mm x 2mm rectangle drawn into the same dc. They should both be nearly the same height. If the text is drawn as 9mm or 10mm (4 or 5 times the size of the rectangle) then it is wrong for this dc.
If it is a bug then it needs to at least be documented. If it is not a bug and for some reason it is proper for the dc to draw text 3 or 4 times as big as it should then that should be documented.   My first assumption is always that I screwed up somewhere and I am using the api wrong. I spent more than a good solid week chasing around trying to figure what I was doing wrong.
So I guess what I expect is for the docs to be appropriately tweaked to reflect this behavior. I agree that changing the behavior here would be a bad idea.
Eddie

**Comment by vadz** [comment]
So, just to confirm: you actually do _not_ expect mapping mode to affect the font size, but it does, is that right?

If so, this is IMO natural behaviour (because mapping mode, a.k.a. scaling, affects everything), i.e. if I use a font to draw something and zoom by a factor of 2, I expect the text to become twice as big. I don't really understand why is this controversial, but please feel free to suggest improvements to the documentation (which is contained in simple text files in `interface/wx/font.h` or `dc.h` or wherever) that you think could help with understanding this.

OTOH if I'm still missing something here, i.e. if you expect something else or if it doesn't behave like described above on all platforms, please let me know.

**Comment by doublemax** [comment]
You may know that internally mapping mode is just a scaling factor, but from a user's point of view, it's not clear at all.

When i switch to use mm units instead of pixel units, why should that have an effect on the font size, especially when it's given in a different unit, like pt? For me that's counter intuitive.

**Comment by EddieM0710** [comment]
As a user… I don’t know or care what the dc does internally in its various mapping modes. 
I set the dc map mode to Metric. The coordinates I use to draw rectangles are in mm. I can draw a rectangle 3mm tall and 10 mm wide. A text with 6 point font should fit nicely within this rectangle. I should not have to scale the font before I use it so that it does fit. 
Zooming is not at issue. I can setuserscale to whatever and the text and rectangle should scale the same. With the right scaling the rectangle could look like it is 6in tall on my display with corresponding size text.
The mapping mode does not change what the point size is. The height of a font point is 1/72in.  So a 6 point font has a defined height in mm. The font size should be treated like any other coordinate passed to the dc. The dc should know what the size a 6 point font should be in mm and draw it correctly. 
 With a metric dc I know to pass metric  coordinates to the interface routines and I’m essentially passing a coordinate when I tell it to draw text with a given point size.
Right now for metric mapping the dc doesn’t get the font size correct and I call that a bug.  I don’t really know anything about what that means internally to the dc or what other mappings are doing.
Eddie

**Comment by vadz** [comment]
For me mapping mode is just a convenience for setting the scaling factor and so I find it very strange that you think about it otherwise but clearly we need to mention this in `SetMapMode()` if this is so controversial.

Does everybody agree that documenting that `SetMapMode()` just changes scaling is all that needs to be done here?

BTW the current documentation has some weird note about X which I don't really understand, so it would still be best to test that this actually works the way I expect it to.

**Comment by EddieM0710** [comment]
You say that mapping mod just changes scaling. Does this exclude Fonts? If so why does the setuserscale affect Fonts? Is that different scaling?

**Comment by vadz** [comment]
I'm sorry, I'm hopelessly confused now, probably because I still didn't receive any answers to [my questions](https://github.com/wxWidgets/wxWidgets/issues/24176#issuecomment-1874094709) and I am trying to deduce them from your comments not answering them directly and apparently am doing it wrongly.

I really need to run the tests myself but my mental picture is that fonts **are** affected by scaling, whether it's logical scale (used by map mode) or user scale. Again, this is what I understood from what you wrote. If this isn't the case, then it would change everything.

**Comment by EddieM0710** [comment]
Sorry I answered the best I could and can’t really talk to any of the internals. I am a total novice here. Long time programmer but not in this kind of stuff. I’m retired and doing this to keep my mind active so this is all fun/learning for me.  Please run the test sample I sent. I think what I’m saying is accurate but may not be using the words right to describe it and may have weird concepts of how things work. I think I know how to make my code work but would like to make it its correct as possible.
Eddie

**Comment by EddieM0710** [comment]
I'm trying to figure out what the mapping mode really is. If i do this:
`    wxPaintDC dc( this );
    DoPrepareDC( dc );

    wxSize PPI = dc.GetPPI( );
    dc.SetMapMode( wxMM_METRIC );
    dc.Clear( );

    dc.SetPen( *wxRED_PEN );
    dc.SetBrush( *wxWHITE_BRUSH );
    dc.DrawRectangle( wxRect( 0, 0, 10 * 25.4, 10 * 25.4 ) );

    dc.SetPen( *wxGREEN_PEN );
    wxRect txtRect = wxRect( 2, 2, 25.4, 25.4 );
    dc.DrawRectangle( txtRect );

    dc.SetMapMode( wxMM_TEXT );
    dc.SetPen( *wxBLUE_PEN );

    wxRect blueRect = wxRect( PPI.x * 2, 2, PPI.x, PPI.y );
    dc.DrawRectangle( blueRect );
`
Shouldn't the blue and green rectangle be the same 1in in size?
I'm wondering if the font is not the problem but the metric mapping scale factor is???
Eddie

**Comment by paulcor** [comment]
>Shouldn't the blue and green rectangle be the same 1in in size?

The `wxMM_METRIC` scale factor was recently fixed for GTK3 by 64d1d27da9. The difference in size between point-sized fonts and other drawn elements is unaffected by that change.

**Comment by EddieM0710** [comment]
But when I run the above code the two rectangles are a substantially different size.  Drawing 1in in metric mapping should be the same as drawing 1in in Text mapping.  If I understand they should be the same size.

In my app I have now changed all the wxMM_METRIC to wxMM_TEXT. On the calls to drawing functions scaled all my metric positions and sizes to pixels, scale from pixels to mm on returns from GetTextExtent, etc.  and removed all scaling of fonts.  My resultant drawn image looks really good. As a bonus the image quality was really bad because of the massive image resizing I was having to do in the Metric mode and now it is really good.
Bottom line is my problem is solved and I'm better off than I was.

On the other hand, I believe there is something wrong with wxWidgets DC metric mapping mode. It is certainly possible that it is in the way I'm trying to use it but I can't figure it out. 

I'm happy but I think wxWidgets has a bug here somewhere. It is ok with me to close this but I really think it should be looked into. I'll be happy to answer questions or try out stuff if it helps. Just let me know.
Eddie

**Comment by EddieM0710** [comment]
In case you are interested this is what I have been working hard to achieve. Here are[][(some results in the github repository](https://github.com/EddieM0710/StampTool/blob/master/sample/)).  there are 2 pdf that I generated from my code which is all in this repository.  But it is not quite ready for prime time.
Eddie
